### PR TITLE
Lru cache integration

### PR DESCRIPTION
**Pull request by PradhayA**
## Description of change

A global (row group, column) Parquet caching structure has been introduced to optimise caching and reduce redundant data loading. Caching of row groups and columns is implemented with an LRU mechanism to remove the least frequently accessed (row group, columns), handling scenarios where certain (row group, columns) are filtered out due to predicate pushdown but unnecessarily brought to the cache by the sequential prefetcher.

The caching mechanism operates on a per-file basis instead of per-file handle and follows the structure of the Parquet file (from the metadata). The global caching structure maintains a single cache for each file, shared across all file handles and threads. This ensures that data and Parquet metadata are loaded and cached only once, regardless of the number of file handles accessing the file.

## Does this change impact existing behavior?

No breaking changes have been introduced, although by default, the Parquet prefetcher is now enabled (which will still work as intended). If a Parquet file or no metadata is detected, it defaults to the old prefetching behaviour. Working on making this a command-line flag to provide more control over the prefetching mechanism.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

**Comment by PradhayA** [comment]
Note: Still working on some small things and a couple of last requests from the previous PR:

- [x] Refactoring and improving metadata reading
- [x] Consolidating and improving how parameters are passed to the underlying stream
- [ ] Adding more tests

File: mountpoint-s3/src/prefetch/lru_cache.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
+use linked_hash_map::LinkedHashMap;
+use std::hash::Hash;
+use tracing::trace;
+
+#[derive(Clone, Hash, Eq, PartialEq, Debug)]
+pub struct CacheKey {
+    pub file_id: String,
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/lru_cache.rs]: Shouldn't this be `ObjectId`?

File: mountpoint-s3/src/prefetch/lru_cache.rs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            memory_usage: 0,
+            memory_limit,
+        }
+    }
+
+    pub fn add_entry(&mut self, key: CacheKey, size: usize) -> Vec<CacheKey> {
+        trace!("Adding entry: {:?}", key);
+
+        let mut evicted = Vec::new();
+        // If the entry already exists, update its size and move it to the back (most recently used)
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/lru_cache.rs:34]: You should be able to use `self.entries.insert` here.

File: mountpoint-s3/src/prefetch/parquet_stream.rs [hunk 3]
```diff
@@ -132,39 +172,98 @@ where
     }
 }
 
-/// Returns the raw metadata range if available
-async fn get_raw_metadata_range(raw_metadata: &RawMetadataRef) -> Option<Range<u64>> {
-    raw_metadata.read().await.as_ref().map(|m| m.range.clone())
+/// Default spawn object request as in default_stream
+async fn spawn_default_get_object_request<Client>(
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/parquet_stream.rs:180]: Is this a straight copy from `part_stream`? Can't you just reuse it?

File: mountpoint-s3/src/prefetch/parquet_prefetch.rs [hunk 4]
```diff
@@ -86,83 +90,72 @@ pub async fn read_parquet_metadata<Client>(
...
+        bucket,
+        key,
+        if_match,
+        raw_metadata_range_start..raw_metadata_range_end,
+    )
+    .await
+    {
+        Ok(bytes) => bytes,
+        Err(_) => return Err(PrefetchReadError::MetadataParsingFailed),
+    };
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/parquet_prefetch.rs]:
```suggestion
    let raw_metadata = fetch_object_part(
        &client,
        bucket,
        key,
        if_match,
        raw_metadata_range_start..raw_metadata_range_end,
    )
    .await
    .map_err(|_| PrefetchReadError::MetadataParsingFailed)?;
```

File: mountpoint-s3/src/prefetch.rs [hunk 5]
```diff
@@ -105,8 +108,10 @@ where
 }
 
 pub type ParquetPrefetcher<Runtime> = Prefetcher<ParquetPartStream<Runtime>>;
-pub type MetadataRef = Arc<AsyncRwLock<Option<ParsedMetadata>>>;
+pub type RowgroupColRanges = HashMap<(RowGroupIndex, ColumnIndex), Range<u64>>;
+pub type MetadataRef = Arc<AsyncRwLock<Option<(ParsedMetadata, RowgroupColRanges)>>>;
 pub type RawMetadataRef = Arc<AsyncRwLock<Option<RawMetadata>>>;
+type CachedRef = Arc<DashMap<ObjectId, (RawMetadataRef, MetadataRef, InMemoryCacheRef)>>;
```
Comment (passaro) [inline mountpoint-s3/src/prefetch.rs]: All these type aliases should really be refactored into structs. But besides that, it is not clear to me what is actually mutable (and when). For example, can the metadata for an `ObjectId` change?

File: mountpoint-s3/src/prefetch/parquet_stream.rs [hunk 6]
```diff
@@ -213,32 +330,67 @@ where
...
+
+                for (row_group_col, _col_range) in new_cols {
+                    let col_cache = cache.entry(row_group_col).or_insert_with(BTreeMap::new);
+                    if let Err(e) = merge_ranges(
+                        col_cache,
+                        part_range.clone(),
+                        part.get_checksummed_bytes().clone(),
+                        metadata
+                            .as_ref()
+                            .unwrap()
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/parquet_stream.rs]: If we can safely `unwrap` here, why are we passing an `Option` in the first place?

File: mountpoint-s3/src/prefetch/parquet_stream.rs [hunk 7]
```diff
@@ -213,32 +330,67 @@ where
...
+                            .1
+                            .get(&row_group_col)
+                            .expect("Expected a range for this row group and column"),
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/parquet_stream.rs]: Why?

**Pull request merged**